hector_math/test/test_ring_buffer.cpp

```diff
@@ -0,0 +1,162 @@
+
+
+using namespace hector_math;
+template<typename Scalar>
+class RingBufferTest : public testing::Test
+{
+};
+
+
+typedef testing::Types<float, double> Implementations;
+
+TYPED_TEST_CASE( RingBufferTest, Implementations );
+
+TYPED_TEST( RingBufferTest, basic ) {
+  using Scalar = TypeParam;
+  constexpr size_t max_size = 50;
+  RingBuffer<Scalar,max_size> ringBuffer;
+  ASSERT_FALSE(ringBuffer.full());
+  ASSERT_TRUE(ringBuffer.empty());
+  ringBuffer.push_back(Scalar(0));
+  ASSERT_FALSE(ringBuffer.empty());
+  ASSERT_EQ(ringBuffer.size(),1);
+  for(size_t i=1;i<max_size-1;i++){
+    ringBuffer.push_back(i);
+    ASSERT_EQ(i+1,ringBuffer.size());
+    ASSERT_FALSE(ringBuffer.full());
+    ASSERT_FALSE(ringBuffer.empty());
+  }
+  ringBuffer.push_back(49);
+  ASSERT_FALSE(ringBuffer.empty());
+  ASSERT_TRUE(ringBuffer.full());
+
+  ASSERT_EQ(ringBuffer.read_and_pop_front(),0);
+  ASSERT_FALSE(ringBuffer.empty());
+  ASSERT_FALSE(ringBuffer.full());
+  ASSERT_EQ(1,ringBuffer.read_and_pop_front());
+  ringBuffer.push_back(50);
+  ASSERT_EQ(ringBuffer.size(),49);
+  // CLEAR
+  ringBuffer.clear();
+  ASSERT_TRUE(ringBuffer.empty());
+  ASSERT_FALSE(ringBuffer.full());
+  ASSERT_EQ(ringBuffer.size(),0);
+}
+
+TYPED_TEST(RingBufferTest, iterators){
+  using Scalar = TypeParam;
+  constexpr size_t max_size = 50;
+  RingBuffer<Scalar,max_size> ringBuffer;
+  // FILL RING BUFFER PARTIAL
+  for(int i=0; i<25;i++){
+    ringBuffer.push_back(i);
+  }
+  ASSERT_EQ(ringBuffer.size(),25);
+  int index=0;
+  for(auto elm:ringBuffer){
+    ASSERT_EQ(index++,elm);
+
+  }
+  ASSERT_EQ(index, ringBuffer.size());
+  //FILL AGAIN AND OVERWRITE FIRST ELEMENTS
+  for(int i=0; i<max_size;i++){
+    ringBuffer.push_back(i);
+  }
+  ASSERT_EQ(ringBuffer.size(),max_size);
+  ASSERT_TRUE(ringBuffer.full());
+  index=0;
+  for(auto elm:ringBuffer){
+    ASSERT_EQ(index++,elm);
+  }
+  ASSERT_EQ(index, ringBuffer.size());
+
+  //algorithms
+  // FOR_EACH
+  auto replace_with_3 = [](Scalar& n) { n=3; };
+  std::for_each(ringBuffer.begin(),ringBuffer.end(),replace_with_3);
+  for(auto elm:ringBuffer){
+    ASSERT_EQ(3,elm);
+  }
+  //FILL AGAIN AND OVERWRITE FIRST ELEMENTS
+  for(int i=0; i<max_size;i++){
+    ringBuffer.push_back(i);
+  }
+  // FIND
+  auto x = std::find(ringBuffer.begin(),ringBuffer.end(),7);
+  ASSERT_EQ(*x,7);
+}
+
+TYPED_TEST( RingBufferTest, const_iterator )
+{
+  using Scalar = TypeParam;
+  constexpr size_t max_size = 50;
+  RingBuffer<Scalar,max_size> ringBuffer;
+  // FILL RING BUFFER PARTIAL
+  for(int i=0; i<25;i++){
+    ringBuffer.push_back(i);
+  }
+  ASSERT_EQ(ringBuffer.size(),25);
+  // Normal iterator can read and write
+  auto it = ringBuffer.begin();
+  ASSERT_NO_THROW(*it=-1);
+  ASSERT_EQ(*ringBuffer.begin(), -1);
+  ASSERT_NO_THROW(*it=0);
+  //CONST ITERATOR can oly read
+  int index = 0;
+  for(auto const_it=ringBuffer.cbegin();const_it!=ringBuffer.cend();const_it++){
+    ASSERT_EQ(*const_it,index++);
+  }
```

Comment:
The concept of const is compile-time only so I'm not sure what you want to test here ^^